presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -153,7 +155,7 @@
     private final boolean isDescribe;
 
     private final IdentityHashMap<FunctionCall, Signature> resolvedFunctions = new IdentityHashMap<>();
-    private final Set<SubqueryExpression> scalarSubqueries = newIdentityHashSet();
+    private final Set<Reference<SubqueryExpression>> scalarSubqueries = new LinkedHashSet<>();
```

Comment:
A prominent source of non-determinisms are HashMaps and HashSets, which have a non-deterministic iteration order. This particular `identityHashSet` (replaced) was breaking TPC-DS Q6. I'd like to replace all of the `IdentityHashSet`-s and `IdentityHashMap`-s with wrapper Sets/Maps handling the wrapping and unwrapping of elements into Guava's [Equivalence.Wrapper<T>](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Equivalence.Wrapper.html). Otherwise, all usages of those identity-collections would have to be cluttered with manual wrapping and unwrapping. I'll provide an implementation of the collection soon for further discussion. For now, please let me know if the general idea sounds good :)